toolchain/lower/testdata/interop/cpp/virtual_base.carbon

```diff
@@ -0,0 +1,195 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/int.carbon
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/lower/testdata/interop/cpp/virtual_base.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/lower/testdata/interop/cpp/virtual_base.carbon
+
+// --- diamond.h
+
+struct A {
+  A();
+  int a;
+};
+
+struct B : virtual A {
+  B();
+  int b;
+};
+
+struct C : virtual A {
+  C();
+  int c;
+};
+
+struct D : B, C {
+  int d;
+};
+
+// Layout of D (x86_64):
+//
+// 0:  B | 0: vptr
+// 8:    | 8: int B::b
+// 12: (4 bytes padding)
+// 16: C | 0: vptr
+// 24:   | 8: int C::c
+// 28: int D::d
+// 32: A | 0: int A::a
+// 36: (4 bytes padding)
+// 40: (end of D)
+
+// --- use_diamond.carbon
+
+library "[[@TEST_NAME]]";
+
+import Cpp library "diamond.h";
+
+fn Make() {
+  // This should construct a 40-byte object, using the `...C1` constructor,
+  // which should call the `...C2` constructors for A, B, and C.
```

Comment:
What are "C1" and "C2" here? Or link to an explanation?

Reply:
Expanded comment a bit to try to explain.

I looked for a link with an explanation, but the ABI document doesn't really have a good place that explains this :(